{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5aff4f92-7e85-4cf6-86c3-4faa448a37d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "from Bio import SeqIO\n",
    "import gzip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "98af8328-bc86-4075-8ecb-5c97b97da9ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# snakemake_path = '/mnt/c/Files/University/lab/snakemake-WuMV-6'\n",
    "# base_vcf_dir = \"/mnt/c/Files/University/lab/snakemake-WuMV-6/vcf\"  \n",
    "# reference_file = \"/mnt/c/Files/University/lab/snakemake-WuMV-6/references/wmv6-cms001_028_alco-ref.fasta\"  \n",
    "# output_path = \"/mnt/c/Files/University/lab/snakemake-WuMV-6/output\"  \n",
    "\n",
    "# # os.chdir(vcf_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "85a6a207-8afa-45be-84c7-8c880f6d2780",
   "metadata": {},
   "outputs": [],
   "source": [
    "min_depth = 10\n",
    "min_alt_freq = 0.2\n",
    "reference_seqs = {}\n",
    "\n",
    "for record in SeqIO.parse(reference_file, \"fasta\"):\n",
    "    reference_seqs[record.id] = str(record.seq)\n",
    "EXPECTED_SEGMENTS = len(reference_seqs) ## number of segments\n",
    "\n",
    "for sample_name in os.listdir(base_vcf_dir):\n",
    "    sample_dir = os.path.join(base_vcf_dir, sample_name)\n",
    "\n",
    "    if not os.path.isdir(sample_dir):\n",
    "        continue\n",
    "    vcf_file = os.path.join(sample_dir, f\"{sample_name}.vcf.gz\")\n",
    "    if not os.path.isfile(vcf_file):\n",
    "        continue\n",
    "\n",
    "    vcf_data = {}\n",
    "    mixed_positions = []\n",
    "\n",
    "    with gzip.open(vcf_file, \"rt\") as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"#\"): ## skip header lines\n",
    "                continue\n",
    "\n",
    "            fields = line.rstrip().split(\"\\t\") ## split vcf line to columns, extract info\n",
    "            chrom = fields[0]\n",
    "            pos = int(fields[1])\n",
    "            ref_nt = fields[3]\n",
    "            alt_nt = fields[4].split(\",\")\n",
    "            format_keys = fields[8].split(\":\")\n",
    "            sample_values = fields[9].split(\":\")\n",
    "            fmt = dict(zip(format_keys, sample_values)) ## parse the FORMAT column and save it to a dictionary\n",
    "\n",
    "            ad = list(map(int, fmt[\"AD\"].split(\",\"))) ## split allele depth info to integers\n",
    "            alleles = [ref_nt] + alt_nt ## get list of all alleles in the site\n",
    "            allele_counts = dict(zip(alleles, ad)) ## make a dict with alleles and their read coutns\n",
    "            # print(allele_counts)\n",
    "\n",
    "            depth = sum(allele_counts.values()) ## calculate total coverage across all alleles\n",
    "            low_depth = depth < min_depth ## get sites with low depth coverage\n",
    "\n",
    "            allele_freqs = {}\n",
    "            high_cov_alleles = {}\n",
    "\n",
    "            if not low_depth:\n",
    "                for allele, count in allele_counts.items():\n",
    "                    if count > 0:\n",
    "                        freq = count / depth ## calculate frequencies of alleles\n",
    "                        allele_freqs[allele] = freq\n",
    "                        if freq >= min_alt_freq: ## if frequency is higher or equal than treshold, keep the allele\n",
    "                            high_cov_alleles[allele] = freq\n",
    "\n",
    "            if chrom not in vcf_data:\n",
    "                vcf_data[chrom] = {}\n",
    "\n",
    "            vcf_data[chrom][pos] = { ## store all calculated info of vcfs in one dict\n",
    "                \"ref\": ref_nt,\n",
    "                \"depth\": depth,\n",
    "                \"alleles\": allele_counts,\n",
    "                \"high\": high_cov_alleles,\n",
    "                \"low_depth\": low_depth\n",
    "            }\n",
    "            # print(vcf_data)\n",
    "\n",
    "            if len(high_cov_alleles) > 1: ## if there is more than 1 allele of high coverage, save it \n",
    "                mixed_positions.append((chrom, pos, ref_nt, depth, high_cov_alleles))\n",
    "\n",
    "    if len(vcf_data) != EXPECTED_SEGMENTS: ## skip sample if there are segments missing\n",
    "        continue\n",
    "\n",
    "## write a report\n",
    "    report_file = os.path.join(output_path, f\"{sample_name}_report.tsv\")\n",
    "\n",
    "    with open(report_file, \"w\", newline=\"\") as f:\n",
    "        writer = csv.writer(f, delimiter=\"\\t\")\n",
    "\n",
    "        writer.writerow([\"sample\", \"segment\", \"position\", \"depth\", \"reference\", \"allele:count:frequency\"]) ## header row\n",
    "\n",
    "        for segment, pos, ref, depth, alleles in mixed_positions:\n",
    "            allele_strs = []\n",
    "            for allele, freq in alleles.items():\n",
    "                count = int(freq * depth)\n",
    "                allele_strs.append(f\"{allele}:{count}:{freq:.4f}\")\n",
    "                # print(allele_strs)\n",
    "                \n",
    "            writer.writerow([sample_name,segment,pos,depth,ref,\",\".join(allele_strs)])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "phylogeo",
   "language": "python",
   "name": "phylogeo"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
